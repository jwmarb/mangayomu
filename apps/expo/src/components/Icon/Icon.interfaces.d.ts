import FontAwesomeIconNames from 'react-native-vector-icons/glyphmaps/FontAwesome5Free.json';
import AntDesignIconNames from 'react-native-vector-icons/glyphmaps/AntDesign.json';
import MaterialCommunityIconNames from 'react-native-vector-icons/glyphmaps/MaterialCommunityIcons.json';
import FoundationIconNames from 'react-native-vector-icons/glyphmaps/Foundation.json';
import FeatherIconNames from 'react-native-vector-icons/glyphmaps/Feather.json';
import OctIconNames from 'react-native-vector-icons/glyphmaps/Octicons.json';
import { AppColors } from '@theme/Color/Color.interfaces';
import { ColorSchemeName } from 'react-native';

export type IconPack = 'FontAwesome5' | 'MaterialCommunityIcons' | 'AntDesign' | 'Foundation' | 'Feather' | 'Octicons';

export type IconProps<T extends IconPack> = {
  bundle: T;
  name: T extends 'FontAwesome5'
    ? keyof typeof FontAwesomeIconNames
    : T extends 'MaterialCommunityIcons'
    ? keyof typeof MaterialCommunityIconNames
    : T extends 'Foundation'
    ? keyof typeof FoundationIconNames
    : T extends 'Feather'
    ? keyof typeof FeatherIconNames
    : T extends 'AntDesign'
    ? keyof typeof AntDesignIconNames
    : T extends 'Octicons'
    ? keyof typeof OctIconNames
    : '';
  color?: AppColors;
  lockTheme?: ColorSchemeName;
  size?: 'small' | 'large' | 'medium' | 'tab' | 'checkbox';
};

export type IconBaseProps<
  T extends typeof FontAwesome5 | typeof MaterialCommunityIcons | typeof AntDesign | typeof Feather | typeof Foundation
> = React.ComponentProps<T>;
