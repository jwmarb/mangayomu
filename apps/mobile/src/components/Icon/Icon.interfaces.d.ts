import MaterialCommunityIconNames from 'react-native-vector-icons/glyphmaps/MaterialCommunityIcons.json';

import React from 'react';
import { TextProps } from '@components/Text/Text.interfaces';

export interface IconProps
  extends React.PropsWithChildren,
    Pick<TextProps, 'color' | 'variant' | 'bold' | 'italic' | 'underline'> {
  name: keyof typeof MaterialCommunityIconNames;
  size?: number;
}
