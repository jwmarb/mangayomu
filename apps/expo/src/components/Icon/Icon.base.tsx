import FontAwesome5 from 'react-native-vector-icons/FontAwesome5';
import MaterialCommunityIcons from 'react-native-vector-icons/MaterialCommunityIcons';
import AntDesign from 'react-native-vector-icons/AntDesign';
import Feather from 'react-native-vector-icons/Feather';
import Foundation from 'react-native-vector-icons/Foundation';
import OctIcons from 'react-native-vector-icons/Octicons';
import styled, { css } from 'styled-components/native';
import { IconBaseProps } from '@components/Icon/Icon.interfaces';

export const FontAwesome5Base = styled(FontAwesome5)<IconBaseProps<typeof FontAwesome5>>``;

export const AntDesignBase = styled(AntDesign)<IconBaseProps<typeof AntDesign>>``;

export const MaterialCommunityIconsBase = styled(MaterialCommunityIcons)<
  IconBaseProps<typeof MaterialCommunityIcons>
>``;

export const FoundationBase = styled(Foundation)<IconBaseProps<typeof Foundation>>``;

export const FeatherBase = styled(Feather)<IconBaseProps<typeof Feather>>``;

export const OcticonsBase = styled(OctIcons)<IconBaseProps<typeof OctIcons>>``;
