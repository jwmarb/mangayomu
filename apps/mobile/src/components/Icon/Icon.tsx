import styled, { css } from '@emotion/native';
import React from 'react';
import { IconProps, ImageIconProps } from './';
import MaterialCommunityIcons from 'react-native-vector-icons/MaterialCommunityIcons';
import { moderateScale, ScaledSheet } from 'react-native-size-matters';
import MaterialCommunityIconNames from 'react-native-vector-icons/glyphmaps/MaterialCommunityIcons.json';
import Animated from 'react-native-reanimated';
import { getOrUseCustomColor } from '@components/Text';
import { Image } from 'react-native';
import ImprovedImage from '@components/ImprovedImage';

const styles = ScaledSheet.create({
  imageIcon: {
    borderRadius: '4@ms',
  },
});

const AnimatedIcon = Animated.createAnimatedComponent(MaterialCommunityIcons);

const Icon: React.FC<IconProps | ImageIconProps> = (props) => {
  if (props.name == null) throw Error('name is required in Icon');
  if (props.type === 'font' && props.name in MaterialCommunityIconNames)
    return (
      <IconFromFont
        as={props.animated ? AnimatedIcon : undefined}
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        {...(props as unknown as any)}
      />
    );
  const { size = moderateScale(32) } = props;
  return (
    <ImprovedImage // ImprovedImage
      source={{ uri: props.name }}
      style={[styles.imageIcon, { width: size, height: size }]}
    />
  );
};

const IconFromFont = styled(MaterialCommunityIcons)<IconProps>`
  ${(props) => {
    const {
      theme,
      color = 'textPrimary',
      bold,
      italic,
      underline,
      variant = 'body',
      size,
    } = props;
    return css`
      ${variant !== 'inherit' && theme.typography[variant]};
      ${variant !== 'inherit' &&
      color !== 'inherit' &&
      `color: ${getOrUseCustomColor(theme, color)}`};
      ${size != null && 'font-size: ' + size + 'px'};
      ${bold && 'font-weight: bold'};
      ${italic && 'font-style: italic'};
      ${underline && 'text-decoration-line: underline'};
    `;
  }}
`;
export default Icon;
