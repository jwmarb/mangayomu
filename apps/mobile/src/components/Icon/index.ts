import Icon from './Icon';
export default Icon;
import MaterialCommunityIconNames from 'react-native-vector-icons/glyphmaps/MaterialCommunityIcons.json';
import React from 'react';
import { TextProps } from '@components/Text';

interface AbstractIconProps<T>
  extends React.PropsWithChildren,
    Pick<TextProps, 'color' | 'variant' | 'bold' | 'italic' | 'underline'> {
  type?: T;
}

export interface IconProps extends AbstractIconProps<'font'>, TextProps {
  name?: keyof typeof MaterialCommunityIconNames;
  size?: number;
  animated?: boolean;
}

export interface ImageIconProps extends AbstractIconProps<'image'> {
  name?: string;
  size?: number;
}
