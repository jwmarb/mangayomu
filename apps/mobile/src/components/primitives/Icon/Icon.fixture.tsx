import { SafeAreaView } from 'react-native-safe-area-context';
import { View } from 'react-native';
import { useSelect } from 'react-cosmos/client';
import MaterialCommunityIcons from 'react-native-vector-icons/glyphmaps/MaterialCommunityIcons.json';
import Icon from '@/components/primitives/Icon';
import { createStyles } from '@/utils/theme';
import useStyles from '@/hooks/useStyles';
import Text from '@/components/primitives/Text';
import useModeSelect from '@/hooks/useModeSelect';
import { TEXT_COLORS, TextColors } from '@/components/primitives/types';

const styles = createStyles((theme) => ({
  container: {
    flex: 1,
    backgroundColor: theme.palette.background.default,
    justifyContent: 'center',
  },
  textContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    gap: theme.style.size.s,
  },
}));

export default function IconFixture() {
  const contrast = useModeSelect();
  const style = useStyles(styles, contrast);
  const [color] = useSelect('color', {
    options: TEXT_COLORS as unknown as string[],
    defaultValue: 'textPrimary',
  }) as unknown as [TextColors];
  const [name] = useSelect('icon-name', {
    options: Object.keys(MaterialCommunityIcons),
    defaultValue: 'ab-testing',
  }) as unknown as [keyof typeof MaterialCommunityIcons];
  return (
    <SafeAreaView style={style.container}>
      <View style={style.textContainer}>
        <Icon type="icon" name={name} size="small" color={color} />
        <Icon type="icon" name={name} size="medium" color={color} />
        <Icon type="icon" name={name} size="large" color={color} />
      </View>
      <View style={style.textContainer}>
        <Text variant="body2" color={color}>
          body2
        </Text>
        <Text color={color}>body1</Text>
        <Text variant="h3" color={color}>
          h3
        </Text>
      </View>
    </SafeAreaView>
  );
}
