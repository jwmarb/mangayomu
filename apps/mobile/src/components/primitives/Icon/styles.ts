import { StyleSheet, TextStyle } from 'react-native';
import { IconSizes } from '@/components/primitives/types';
import { IconProps } from '@/components/primitives/Icon/Icon';

export const sizes = (type: Pick<IconProps, 'type'>['type']) =>
  StyleSheet.create<Record<IconSizes, TextStyle>>({
    large:
      type === 'icon'
        ? {
            fontSize: 32,
          }
        : {
            width: 32,
            height: 32,
          },
    medium:
      type === 'icon'
        ? {
            fontSize: 24,
          }
        : {
            width: 24,
            height: 24,
          },
    small:
      type === 'icon'
        ? {
            fontSize: 16,
          }
        : {
            width: 16,
            height: 16,
          },
  });
