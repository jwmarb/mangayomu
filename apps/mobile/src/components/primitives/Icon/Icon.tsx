import MaterialCommunityIconNames from 'react-native-vector-icons/glyphmaps/MaterialCommunityIcons.json';
import MaterialCommunityIcons from 'react-native-vector-icons/MaterialCommunityIcons';
import {
  Image,
  ImageStyle,
  StyleProp,
  TextProps,
  TextStyle,
} from 'react-native';
import { sizes } from '@/components/primitives/Icon/styles';
import { IconSizes, TextColors } from '@/components/primitives/types';
import { colors } from '@/components/primitives/Text/styles';
import useStyles from '@/hooks/useStyles';
import useContrast from '@/hooks/useContrast';

export type IconProps = BaseProps & (FontIconProps | ImageIconProps);

type BaseProps = Pick<TextProps, 'style'> & {
  color?: TextColors;
  size?: 'large' | 'medium' | 'small';
  contrast?: boolean;
  testID?: string;
};
type FontIconProps = {
  type: 'icon';
  name: keyof typeof MaterialCommunityIconNames;
};

type ImageIconProps = {
  type: 'image';
  uri: string;
};

export const composedSizes = (['icon', 'image'] as const).reduce(
  (prev, curr) => {
    prev[curr] = sizes(curr);
    return prev;
  },
  {} as Record<Pick<IconProps, 'type'>['type'], Record<IconSizes, TextStyle>>,
);

export default function Icon(props: IconProps) {
  const {
    color: colorProp = 'textPrimary',
    size: sizeProp = 'medium',
    contrast: contrastProp,
    style: styleProp,
    ...rest
  } = props;
  const contrast = useContrast(contrastProp);
  const size = composedSizes[rest.type][sizeProp];
  const color = useStyles(colors, contrast)[colorProp];
  const style =
    rest.type === 'icon' ? [size, color, styleProp] : [size, styleProp];
  switch (rest.type) {
    case 'icon':
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      return (
        <MaterialCommunityIcons
          {...rest}
          name={rest.name}
          style={style as any}
        />
      );
    case 'image':
      return (
        <Image
          {...rest}
          src={rest.uri}
          style={style as StyleProp<ImageStyle>[]}
        />
      );
  }
}
