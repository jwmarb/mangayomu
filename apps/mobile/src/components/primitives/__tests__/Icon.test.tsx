import { render } from '@testing-library/react-native';
import Icon, { composedSizes } from '@/components/primitives/Icon';

jest.mock('react-native-vector-icons/MaterialCommunityIcons', () => ({
  __esModule: true,
  default: (props: Record<string, unknown>) => {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const { Text } = require('react-native');
    return <Text {...props}>icon</Text>;
  },
}));

test('renders properly', () => {
  let tree = render(<Icon type="icon" name="ab-testing" />);
  expect(tree).toMatchSnapshot();
  tree = render(<Icon type="image" uri="uri" />);
  expect(tree).toMatchSnapshot();
});

test('styles properly', () => {
  let tree = render(
    <>
      <Icon type="icon" name="ab-testing" size="small" testID="small" />
      <Icon type="icon" name="ab-testing" size="medium" testID="medium" />
      <Icon type="icon" name="ab-testing" size="large" testID="large" />
    </>,
  );

  expect(tree.getByTestId('small')).toHaveStyle(composedSizes.icon.small);
  expect(tree.getByTestId('medium')).toHaveStyle(composedSizes.icon.medium);
  expect(tree.getByTestId('large')).toHaveStyle(composedSizes.icon.large);

  tree = render(
    <>
      <Icon type="image" uri="uri" size="small" testID="small" />
      <Icon type="image" uri="uri" size="medium" testID="medium" />
      <Icon type="image" uri="uri" size="large" testID="large" />
    </>,
  );

  expect(tree.getByTestId('small')).toHaveStyle(composedSizes.image.small);
  expect(tree.getByTestId('medium')).toHaveStyle(composedSizes.image.medium);
  expect(tree.getByTestId('large')).toHaveStyle(composedSizes.image.large);
});
