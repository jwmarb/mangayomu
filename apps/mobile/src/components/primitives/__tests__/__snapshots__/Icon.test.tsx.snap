// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders properly 1`] = `
<Text
  name="ab-testing"
  style={
    [
      {
        "fontSize": 24,
      },
      {
        "color": "rgba(0, 0, 0, 0.87)",
      },
      undefined,
    ]
  }
  type="icon"
>
  icon
</Text>
`;

exports[`renders properly 2`] = `
<Image
  src="uri"
  style={
    [
      {
        "height": 24,
        "width": 24,
      },
      undefined,
    ]
  }
  type="image"
  uri="uri"
/>
`;
